packages/network-enablement-controller/src/NetworkEnablementController.ts

```diff
@@ -0,0 +1,372 @@
+import { BaseController } from '@metamask/base-controller';
+import type {
+  RestrictedMessenger,
+  ControllerGetStateAction,
+  ControllerStateChangeEvent,
+} from '@metamask/base-controller';
+import { BuiltInNetworkName, ChainId, toHex } from '@metamask/controller-utils';
+import { SolScope } from '@metamask/keyring-api';
+import { toEvmCaipChainId } from '@metamask/multichain-network-controller';
+import type {
+  NetworkConfiguration,
+  NetworkControllerNetworkAddedEvent,
+  NetworkControllerNetworkRemovedEvent,
+  NetworkControllerStateChangeEvent,
+} from '@metamask/network-controller';
+import type { CaipChainId, CaipNamespace, Hex } from '@metamask/utils';
+import {
+  isCaipChainId,
+  isHexString,
+  KnownCaipNamespace,
+  parseCaipChainId,
+} from '@metamask/utils';
+
+// Unique name for the controller
+const controllerName = 'NetworkEnablementController';
+
+/**
+ * Information about an ordered network.
+ */
+export type NetworksInfo = {
+  networkId: CaipChainId; // The network's chain id
+};
+
+// State shape for NetworkEnablementController
+export type NetworkEnablementControllerState = {
+  enabledNetworkMap: Record<CaipNamespace, Record<string, boolean>>;
+};
+
+export type NetworkEnablementControllerGetStateAction =
+  ControllerGetStateAction<
+    typeof controllerName,
+    NetworkEnablementControllerState
+  >;
+
+export type NetworkEnablementControllerSetEnabledNetworksAction = {
+  type: `${typeof controllerName}:setEnabledNetworks`;
+  handler: NetworkEnablementController['setEnabledNetworks'];
+};
+
+export type NetworkEnablementControllerDisableNetworkAction = {
+  type: `${typeof controllerName}:disableNetwork`;
+  handler: NetworkEnablementController['disableNetwork'];
+};
+
+export type NetworkEnablementControllerIsNetworkEnabledAction = {
+  type: `${typeof controllerName}:isNetworkEnabled`;
+  handler: NetworkEnablementController['isNetworkEnabled'];
+};
+
+export type AllowedActions =
+  | NetworkEnablementControllerGetStateAction
+  | NetworkEnablementControllerSetEnabledNetworksAction
+  | NetworkEnablementControllerDisableNetworkAction
+  | NetworkEnablementControllerIsNetworkEnabledAction;
+
+export type NetworkEnablementControllerActions =
+  | NetworkEnablementControllerGetStateAction
+  | NetworkEnablementControllerSetEnabledNetworksAction
+  | NetworkEnablementControllerDisableNetworkAction
+  | NetworkEnablementControllerIsNetworkEnabledAction;
+
+export type NetworkEnablementControllerStateChangeEvent =
+  ControllerStateChangeEvent<
+    typeof controllerName,
+    NetworkEnablementControllerState
+  >;
+
+export type NetworkEnablementControllerEvents =
+  NetworkEnablementControllerStateChangeEvent;
+
+export type AllowedEvents =
```

Comment:
Nit: this type is not supposed to be exported
```suggestion
type AllowedEvents =